### PR TITLE
Recipe AddSerialAnnotationToserialVersionUID

### PR DESCRIPTION
**Pull request by jbessels**
## What's changed?
Added recipe AddSerialAnnotationToserialVersionUID.  Ran recipe AddSerialVersionUidToSerializable on the firm repo's and quite a few files were changed. Since Java 14 the @Serial annotation exists. The current recipe does not add that annotation. 

## What's your motivation?

This is my first recipe ever and a learning experience to get the ball rolling. That is why I created a new recipe instead of creating a feature request ticket that asks to add that @Serial annotation in the AddSerialAnnotationToserialVersionUID recipe. My recipe will only add a @Serial but will not change/fix the existing code. The correct format is
`private static final long serialVersionUID = 1;`
Sometimes private and/or static is missing, it works but not according to the standard. Such cases will not be corrected by my recipe. For that, there is recipe AddSerialVersionUidToSerializable if a user chooses to use that. 

## Anything in particular you'd like reviewers to focus on?
As its my first recipe and the Moderne API is new to me I will make mistakes. No need to be gentle, just tell me what I'm doing wrong and/or what I'm doing correct (if so). I need the recipe to be elegant, fast and applying the best practices of Moderne. 

## Anyone you would like to review specifically?
@timtebeek offered to contribute/help me with this recipe.

## Have you considered any alternatives or workarounds?
My first approach was based upon AddSerialAnnotationToserialVersionUID and the main logic was in the visitClassDeclaration() method. Turned out that using that approach its not possible or not easy/elegant to put a @Serial above a serialVersionUID line. It was always put above the class. @timtebeek suggested to use **visitVariableDeclarations()**  instead as it would it easier and being able to insert that @Serial above the mentioned line.

## Any additional context
None

### Checklist
- [x] I've added unit tests to cover both positive and negative cases
- [x] I've read and applied the [recipe conventions and best practices](https://docs.openrewrite.org/authoring-recipes/recipe-conventions-and-best-practices)
- [x] I've used the IntelliJ IDEA auto-formatter on affected files

### CODE REVIEW

File: src/main/java/org/openrewrite/staticanalysis/AddSerialAnnotationToserialVersionUID.java [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+ * <p>
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.openrewrite.staticanalysis;
+
+import org.jetbrains.annotations.NotNull;
```
Comment (timtebeek) [inline src/main/java/org/openrewrite/staticanalysis/AddSerialAnnotationToserialVersionUID.java:18]: No need to use this annotation; we have that on the package by default with our own internal annotation.
```suggestion
```

File: src/main/java/org/openrewrite/staticanalysis/AddSerialAnnotationToserialVersionUID.java [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+import java.util.function.UnaryOperator;
+
+public class AddSerialAnnotationToserialVersionUID extends Recipe {
+
+
+
+    @Override
+    @NotNull
+    public String getDisplayName() {
+        System.out.println("Entering getDisplayName");
```
Comment (timtebeek) [inline src/main/java/org/openrewrite/staticanalysis/AddSerialAnnotationToserialVersionUID.java:44]:
```suggestion
```

File: src/main/java/org/openrewrite/staticanalysis/AddSerialAnnotationToserialVersionUID.java [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+        return "Add @Serial annotation to serialVersionUID";
+    }
+
+    @Override
+    @NotNull
+    public String getDescription() {
+        System.out.println("Entering getDescription");
```
Comment (timtebeek) [inline src/main/java/org/openrewrite/staticanalysis/AddSerialAnnotationToserialVersionUID.java:51]:
```suggestion
```

File: src/main/java/org/openrewrite/staticanalysis/AddSerialAnnotationToserialVersionUID.java [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+        return "Add a @Serial annotation above a serialVersionUID attribute identifier line.";
+    }
+
+    @Override
+    @NotNull
+    public Duration getEstimatedEffortPerOccurrence() {
+        System.out.println("Entering getEstimatedEffortPerOccurrence");
```
Comment (timtebeek) [inline src/main/java/org/openrewrite/staticanalysis/AddSerialAnnotationToserialVersionUID.java:58]:
```suggestion
```

File: src/test/java/org/openrewrite/staticanalysis/AddSerialAnnotationToserialVersionUIDTest.java [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+                             import java.io.Serializable;
+                             import java.io.Serial;
+                                         
+                             public class Example implements Serializable {
+                                 String var1 = "first variable";
+                                 @Serial
+                                 private static final long serialVersionUID = 1L;
+                                 int var3 = 666;
+                                 }
+                    """
```
Comment (timtebeek) [inline src/test/java/org/openrewrite/staticanalysis/AddSerialAnnotationToserialVersionUIDTest.java:46]: Could you have IntelliJ apply it's formatter? I think the shortcut is `Ctrl + Alt + L` on Windows too.

File: src/test/java/org/openrewrite/staticanalysis/AddSerialAnnotationToserialVersionUIDTest.java [hunk 6]
```diff
@@ -0,0 +1,133 @@
...
+                        int var3 = 666;
+                        String wolvie = "wolverine"; 
+                    }
+                    """
+          )
+        );
+    }
+
+
+    // Borowed from AddSerialVersionUidToSerializableTest of recipe AddSerialVersionUidToSerializableTest
```
Comment (timtebeek) [inline src/test/java/org/openrewrite/staticanalysis/AddSerialAnnotationToserialVersionUIDTest.java:84]:
```suggestion
```

File: src/test/java/org/openrewrite/staticanalysis/AddSerialAnnotationToserialVersionUIDTest.java [hunk 7]
```diff
@@ -0,0 +1,133 @@
...
+                        void doSomething() {
+                            long serialVersionUID = 1L;
+                        }
+                    }
+                    """
+          )
+        );
+    }
+
+    // Borowed from AddSerialVersionUidToSerializableTest of recipe AddSerialVersionUidToSerializableTest
```
Comment (timtebeek) [inline src/test/java/org/openrewrite/staticanalysis/AddSerialAnnotationToserialVersionUIDTest.java:106]:
```suggestion
```

**Review by timtebeek — CHANGES_REQUESTED** [review]
Couple first quick suggestion to help readability and prevent conflicts when we work on the same files.

File: src/test/java/org/openrewrite/staticanalysis/AddSerialAnnotationToserialVersionUIDTest.java [hunk 8]
```diff
@@ -0,0 +1,133 @@
...
+                                 @Serial
+                                 private static final long serialVersionUID = 1L;
+                                 int var3 = 666;
+                                 }
+                    """
+          )
+        );
+    }
+
+    @Disabled
```
Comment (timtebeek) [inline src/test/java/org/openrewrite/staticanalysis/AddSerialAnnotationToserialVersionUIDTest.java:51]:
```suggestion
    @DocumentExample
```

File: src/main/java/org/openrewrite/staticanalysis/AddSerialAnnotationToserialVersionUID.java [hunk 9]
```diff
@@ -0,0 +1,155 @@
...
+                AtomicBoolean needsSerialAnnotation = new AtomicBoolean(false);
+                c = c.withBody(c.getBody().withStatements(ListUtils.map(c.getBody().getStatements(), new UnaryOperator<Statement>() {
+                    @Override
+                    public Statement apply(Statement s) {
+                        if (!(s instanceof J.VariableDeclarations)) {
+                            return s;
+                        }
+                        J.VariableDeclarations varDecls = (J.VariableDeclarations) s;
+                        // Yes I know deprecated: varDecls.getAllAnnotations()
+                        List<J.Annotation> allAnnotations = varDecls.getAllAnnotations();
```
Comment (timtebeek) [inline src/main/java/org/openrewrite/staticanalysis/AddSerialAnnotationToserialVersionUID.java:105]: Indeed elsewhere we've already adopted the annotation service; you'll want to look into that before we merge.

File: src/main/java/org/openrewrite/staticanalysis/AddSerialAnnotationToserialVersionUID.java [hunk 10]
```diff
@@ -0,0 +1,155 @@
...
+    @NotNull
+    public Duration getEstimatedEffortPerOccurrence() {
+        System.out.println("Entering getEstimatedEffortPerOccurrence");
+        return Duration.ofMinutes(1);
+    }
+
+    @Override
+    @NotNull
+    public TreeVisitor<?, ExecutionContext> getVisitor() {
+        return new JavaIsoVisitor<ExecutionContext>() {
```
Comment (timtebeek) [inline src/main/java/org/openrewrite/staticanalysis/AddSerialAnnotationToserialVersionUID.java:65]: We'll want to use a precondtion here to check that we use at least Java 14+

File: src/main/java/org/openrewrite/staticanalysis/AddSerialAnnotationToserialVersionUID.java (hunk 3 above)
Comment (jbessels) [reply src/main/java/org/openrewrite/staticanalysis/AddSerialAnnotationToserialVersionUID.java:51]: In the final code ofc no System.out.println() statement will exist. For now its riddled with the as to see how the flow of a recipe is when called/executed.

File: src/test/java/org/openrewrite/staticanalysis/AddSerialAnnotationToserialVersionUIDTest.java (hunk 5 above)
Comment (jbessels) [reply src/test/java/org/openrewrite/staticanalysis/AddSerialAnnotationToserialVersionUIDTest.java:46]: I did use that but for some reason it garbles things. I borrowed test code from another recipe where it was formatted corrected. When using the formatter it got garbled. Will look into this to get it right.

File: src/test/java/org/openrewrite/staticanalysis/AddSerialAnnotationToserialVersionUIDTest.java (hunk 8 above)
Comment (jbessels) [reply src/test/java/org/openrewrite/staticanalysis/AddSerialAnnotationToserialVersionUIDTest.java:51]: Disabled cause the logic put the @serial in the wrong place. What does the @DocumentExample annotatin do?

File: src/main/java/org/openrewrite/staticanalysis/AddSerialAnnotationToserialVersionUID.java (hunk 9 above)
Comment (jbessels) [reply src/main/java/org/openrewrite/staticanalysis/AddSerialAnnotationToserialVersionUID.java:105]: Will be corrected if I still need it with the new approach visitVariableDeclarations() approach.

File: src/main/java/org/openrewrite/staticanalysis/AddSerialAnnotationToserialVersionUID.java (hunk 10 above)
Comment (jbessels) [reply src/main/java/org/openrewrite/staticanalysis/AddSerialAnnotationToserialVersionUID.java:65]: Indeed. Was on the TODO list.

File: src/test/java/org/openrewrite/staticanalysis/AddSerialAnnotationToserialVersionUIDTest.java (hunk 8 above)
Comment (timtebeek) [reply src/test/java/org/openrewrite/staticanalysis/AddSerialAnnotationToserialVersionUIDTest.java:51]: We use DocumentExample to indicate which test(s) best signifies what change the recipe makes in practice; in that past we featured that in the docs, which we hope to restore again soon.

**Review by timtebeek — APPROVED** [review]
Hi @jbessels ; Thanks for getting this started! Ended up applying the necessary changes to complete this one such that we can close out this old PR. Feel free to look at the changes I've made since to see what you can still learn from seeing this implemented. In particular note how we're no longer overriding `visitClassDeclaration` for the bulk of the work, but instead have pushed that down to `visitVariableDeclarations`.

**Comment by timtebeek** [comment]
- Fixes https://github.com/openrewrite/rewrite-migrate-java/issues/96

**Comment by melloware** [comment]
Should this be part of the Java17 automatic migration?  Or part of the 14 migration that happens as part of the 17 migration Recipe?

**Pull request merged**